docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,372 @@
+    } else {
+      coins[i] = coin
+    }
+  }
+
+  // escrow tokens if the destination chain is the same as the sender's
+  escrowAddress := GetEscrowAddress(sourcePort, sourceChannel)
+
+  // escrow source tokens. It fails if balance insufficient.
+  if err := k.bankKeeper.SendCoins(
+    ctx, sender, escrowAddress, coins,
+  ); err != nil {
+    return err
+  }
+} else {
+  // build the receiving denomination prefix if it's not present
+  prefix = GetDenomPrefix(sourcePort, sourceChannel)
+  for _, coin := range amount {
+    if !strings.HasPrefix(denomTrace.Trace, prefix) {
+      return Wrapf(ErrInvalidDenomForTransfer, "invalid token denom trace prefix: %s", denomTrace.IBCDenom())
+    }
+  }
+// ...
+}
+```
+
+```golang
+// OnRecvPacket
+// ...
+prefix := types.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
+source := strings.HasPrefix(data.DenomTrace.Trace, prefix)
+
+// ...
+// set the value to the lookup table if not stored already
+traceHash := denomTrace.Hash()
+if !k.HasDenomTrace(ctx, traceHash) {
+  k.SetDenomTrace(ctx, traceHash, denomTrace)
+}
+
+if source {
+  // ... (no changes)
+}
+
+// check the denom prefix
+prefix = types.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+coins := make(sdk.Coins, len(data.Amount))
+for i, coin := range data.Amount {
+  if !strings.HasPrefix(denomTrace.Trace, prefix) {
+    return Wrapf(types.ErrInvalidDenomForTransfer, "invalid token denom trace prefix: %s", denomTrace.IBCDenom(),
+    )
+  }
+
+  if err := denomTrace.RemovePrefix(); err != nil {
+    return err
+  }
+
+  // set the new value to the lookup table if not stored already
+  traceHash := denomTrace.Hash()
+  if !k.HasDenomTrace(ctx, traceHash) {
+    k.SetDenomTrace(ctx, traceHash, denomTrace)
+  }
+
+  coins[i] = sdk.NewCoin(denomTrace.IBCDenom(), coin.Amount)
+}
+// ...
+```
+
+### Coin Changes
+
+The coin denomination validation will need to be updated to reflect these changes. In particular, the denomination validation
+function will now accept slash separators (`"/"`) and will bump the maximum character length to 64.
+
+Additional validation logic, such as verifying the kenght of the hash, the  may be added to the bank module in the future if the [custom base denomination validation](https://github.com/cosmos/cosmos-sdk/pull/6755) is integrated into the SDK.
+
+### Positive
+
+- Clearer separation of the source tracing behaviour of the token (transfer prefix) from the original
+  `Coin` denomination
+- Consistent validation of `Coin` fields (i.e no special characters, fixed max length)
+- Cleaner `Coin` and standard denominations for IBC
+- No additional fields to SDK `Coin`
+
+### Negative
+
+- Store each set of tracing denomination identifiers on the `ibc-transfer` module store
+- Clients will have to fetch the base denomination everytime they receive a new relayed fungible token over IBC. This can be mitigated using a map/cache for already seen hashes on the client side.
```

Comment:
listening to events would assist with this as well or querying for a balance with metadata

Reply:
good call!